src/poly/dcrt/matrix/dcrt_poly.rs

```diff
@@ -202,72 +208,83 @@ impl PolyMatrix for DCRTPolyMatrix {
-    //                     .collect::<Vec<_>>()
-    //             })
-    //             .collect::<Vec<_>>()
-    //     };
-    //     matrix.replace_entries(0..nrow, 0..ncol, f);
-    //     matrix
-    // }
-
-    // fn write_to_files<P: AsRef<Path> + Send + Sync>(&self, dir_path: P) {
-    //     let block_size = block_size();
-    //     let (row_offsets, col_offsets) = block_offsets(0..self.nrow, 0..self.ncol);
-    //     parallel_iter!(row_offsets.iter().tuple_windows().collect_vec()).for_each(
-    //         |(cur_block_row_idx, next_block_row_idx)| {
-    //             parallel_iter!(col_offsets.iter().tuple_windows().collect_vec()).for_each(
-    //                 |(cur_block_col_idx, next_block_col_idx)| {
-    //                     let entries = self.block_entries(
-    //                         *cur_block_row_idx..*next_block_row_idx,
-    //                         *cur_block_col_idx..*next_block_col_idx,
-    //                     );
-    //                     let mut path = dir_path.as_ref().to_path_buf();
-    //                     path.push(format!(
-    //                         "{}_{}.{}_{}.{}.matrix",
-    //                         block_size,
-    //                         cur_block_row_idx,
-    //                         next_block_row_idx,
-    //                         cur_block_col_idx,
-    //                         next_block_col_idx
-    //                     ));
-    //                     let entries_bytes: Vec<Vec<Vec<u8>>> = entries
-    //                         .iter()
-    //                         .map(|row| row.iter().map(|poly|
-    // poly.to_compact_bytes()).collect_vec())                         .collect_vec();
-    //                     serde_json::to_writer(
-    //                         std::fs::File::create(&path).unwrap(),
-    //                         &entries_bytes,
-    //                     )
-    //                     .unwrap_or_else(|_| panic!("Failed to write matrix file {:?}", path));
-    //                 },
-    //             );
-    //         },
-    //     );
-    // }
+    fn read_from_files<P: AsRef<Path> + Send + Sync>(
+        params: &<Self::P as Poly>::Params,
+        nrow: usize,
+        ncol: usize,
+        dir_path: P,
+        id: &str,
+    ) -> Self {
+        let block_size = block_size();
+        let mut matrix = Self::new_empty(params, nrow, ncol);
+
+        let (row_offsets, col_offsets) = block_offsets(0..nrow, 0..ncol);
+
+        for (row_start, row_end) in row_offsets.iter().tuple_windows() {
+            for (col_start, col_end) in col_offsets.iter().tuple_windows() {
+                let f = |row_range: Range<usize>, col_range: Range<usize>| -> Vec<Vec<DCRTPoly>> {
+                    let mut path = dir_path.as_ref().to_path_buf();
+                    path.push(format!(
+                        "{}_{}_{}.{}_{}.{}.matrix",
+                        id, block_size, row_start, row_end, col_start, col_end
```

Comment:
The `row_start`, `row_end`, `col_start`, `col_end` seem to be independent of the `row_range` and `col_range` provided for `f`.